### PR TITLE
add nft tile component

### LINKED ISSUE #1791: Create an "NFT tile" component

### PR DESCRIPTION
**Pull request by rafalpawlow**
closes #1791

### CODE REVIEW

File: src/components/NftTileDetails/NftTileDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+      ref={contentRef}
+      loading={loading}
+      onMouseEnter={toggleContentHover}
+      onMouseLeave={toggleContentHover}
+      contextMenuHovered={contextMenuHovered}
+      tileSize={tileSize}
+    >
+      <Header>
+        <StyledAvatarGroup
+          avatarStrokeColor={contentHovered && !contextMenuHovered ? cVar('colorBackground', true) : undefined}
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.tsx]: avatar group has a wrong color stroke: 
image

File: src/components/NftTile/NftTile.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    return () => {
+      clearInterval(timeLeftInterval)
+    }
+  }, [timeLeft])
+
+  const getBottomLeft = useMemo(() => {
+    switch (auction) {
+      case 'none':
+        return <Pill icon={buyNow ? <SvgActionBuyNow /> : <SvgActionNotForSale />} size="medium" />
```
Comment (drillprop) [inline src/components/NftTile/NftTile.tsx]: We should use variant `overlay` for all the pills
image

File: src/components/NftTile/NftTile.tsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+          <PillGroup
+            items={[
+              {
+                icon: <SvgActionAuction />,
+                label: timeLeft
+                  ? calculatedTimeLeft < 60
+                    ? 'Less than a minute'
+                    : formatDurationShort(calculatedTimeLeft, true)
+                  : undefined,
+                variant: timeLeft && calculatedTimeLeft < 3600 ? 'danger' : 'default',
```
Comment (drillprop) [inline src/components/NftTile/NftTile.tsx]: same as above

File: src/components/NftTile/NftTile.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            icon={<SvgActionAuction />}
+            label={
+              timeLeft
+                ? calculatedTimeLeft < 60
+                  ? 'Less than a minute'
+                  : formatDurationShort(calculatedTimeLeft, true)
+                : undefined
+            }
+            size="medium"
+            variant={timeLeft && calculatedTimeLeft < 3600 ? 'danger' : 'default'}
```
Comment (drillprop) [inline src/components/NftTile/NftTile.tsx]: same as above

File: src/components/NftTile/NftTile.tsx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+  return (
+    <Container fullWidth={fullWidth}>
+      <VideoThumbnail
+        loading={loading}
+        thumbnailUrl={thumbnail.thumbnailUrl}
+        clickable={false}
+        slots={{
+          topLeft: views
+            ? { element: <Pill label={formatNumberShort(views)} size="medium" icon={<SvgActionShow />} /> }
```
Comment (drillprop) [inline src/components/NftTile/NftTile.tsx]: same as above

File: src/components/NftTile/NftTile.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            : undefined,
+          bottomLeft: { element: getBottomLeft },
+          bottomRight: duration ? { element: <Pill label={formatDurationShort(duration)} size="medium" /> } : undefined,
```
Comment (drillprop) [inline src/components/NftTile/NftTile.tsx]: same as above

File: src/components/NftTileDetails/NftTileDetails.styles.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  left: 0;
+`
+
+export const Title = styled(Text)`
+  min-height: 48px;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  /* stylelint-disable-next-line value-no-vendor-prefix */
+  display: -webkit-box;
+  -webkit-box-orient: vertical;
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.styles.ts:85]: image
We need to set `word-break` as well.

File: src/components/NftTileDetails/NftTileDetails.styles.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      fill: ${({ secondary }) => (secondary ? cVar('colorText') : cVar('colorTextStrong'))};
+    }
+  }
+`
+
+export const CaptionSkeletonWrapper = styled.div`
+  width: 100%;
+`
+
+export const CaptionSkeleton = styled(SkeletonLoader)`
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.styles.ts]: You could also use regular SkeletonLoader and just pass `bottomSpace` prop:
for example:
```tsx
         <SkeletonLoader height={16} width="100%" bottomSpace={4} />
```

File: src/components/NftTile/NftTile.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  minBid,
+  topBid,
+  timeLeft,
+  role,
+  fullWidth,
+}) => {
+  const [calculatedTimeLeft, setCalculatedTimeLeft] = useState(0)
+
+  useEffect(() => {
+    let timeLeftInterval: Timeout
```
Comment (drillprop) [inline src/components/NftTile/NftTile.tsx]: If I understand this correctly we could get rid of this `let` by changing the condition below i.e
```tsx
if (!timeleft) {
   return
}
// rest of the code
```

File: src/components/NftTileDetails/NftTileDetails.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  argTypes: {
+    title: {
+      type: 'string',
+      defaultValue: 'Did An Alternate Reality Game Gone Wrong Predict QAnon?',
+    },
+    creator: { table: { disable: true } },
+    supporters: { table: { disable: true } },
+    owner: { table: { disable: true } },
+    auction: { control: { type: 'select', options: ['none', 'minBid', 'topBid', 'waiting'] }, defaultValue: 'none' },
+    role: { control: { type: 'radio', options: ['owner', 'viewer'] }, defaultValue: 'owner' },
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.stories.tsx:18]:
```suggestion
    auction: { control: { type: 'select', options: ['none', 'minBid', 'topBid', 'waiting'] } },
    role: { control: { type: 'radio', options: ['owner', 'viewer'] } },
```

File: src/components/NftTileDetails/NftTileDetails.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  args: {
+    buyNow: false,
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.stories.tsx:21]:
```suggestion
    buyNow: false,
    role: 'owner',
    auction: 'none',
```

**Review by drillprop — CHANGES_REQUESTED** [review]

File: src/components/NftTile/NftTile.tsx (hunk 9 above)
Comment (rafalpawlow) [reply src/components/NftTile/NftTile.tsx]: This `let` is for `clearInterval` purpose

Comment (drillprop) [reply src/components/NftTile/NftTile.tsx]: Yeah, I know, but my point was that you could do something like this: 
```tsx
  useEffect(() => {
    if (!timeLeft) {
      return
    }

    setCalculatedTimeLeft(timeLeft)
    const timeLeftInterval = setInterval(() => {
      setCalculatedTimeLeft((prevState) => {
        if (prevState <= 1) {
          clearInterval(timeLeftInterval)
        }
        return --prevState
      })
    }, 1000)

    return () => {
      clearInterval(timeLeftInterval)
    }
  }, [timeLeft])
```

Comment (rafalpawlow) [reply src/components/NftTile/NftTile.tsx]: ah, ok 👍

**Review by drillprop — APPROVED** [review]
LGTM!

**Comment by kdembler** [comment]
Let's get approval of @toiletgranny on this one

**Comment by toiletgranny** [comment]
Rafał, this is amazing! Your attention to detail is seriously taking my breath away, I wasn't able to find a single pixel being off! And the stories you set up in Storybook made checking all of that a breeze, so thank you very much for this!

1. When displaying `topBid` and `minBid` prices, it seems like large numbers or not being abbreviated (eg. in comparison to the "Buy now" price).
2. Speaking about abbreviating large numbers, a while ago we refreshed our internal guidelines on how we want to do that (cc @kdembler). [You can find out all about it over here](https://www.notion.so/Date-time-number-formats-in-Atlas-c5e7c7aec24e4ddfa83104ca4c2e924e), but in a nutshell, you might want to change the decimal separator from a dot `.` to a coma `,`; and use spaces ` ` as the thousands separator if needed in all numbers across NFT tile & NFT details components.
3. I'm looking at the countdown in the bottom left corner of the Video thumbnail component and I don't quite like how the changing digits affect the width of the pill component they're wrapped within. It's not anything you implemented wrong by any means, and I know this is a Pill component that we're not allowed to edit 100% freely, but I'm wondering...  I don't suppose we could add this `font-feature-settings: 'tnum' on, 'lnum' on;` CSS property to the timer in this very pill, could we? What it does is it makes all number characters be of an equal width = no jumping left/right when the countdown is on. Let me know.
4. One adjustment we made with @drillprop to the Video tile component implementation was to have hovering over the Video thumbnail triggering the hover of the details below. Could we make the same sort of adjustment over here as well? The end result would be that when I'm hovering over the video thumbnail, the "NFT details" component below goes into its hover state automatically.
5. Text style used for displaying title of an NFT in the "Small" size does not match the current designs. Right now it's `H200`, but it should be `H300`. This is my fault, as we changed this with @kdembler and @drillprop only last week and I wasn't aware you already implemented this part.

**Review by toiletgranny — CHANGES_REQUESTED** [review]
Rafał, this is amazing! Your attention to detail is seriously taking my breath away, I wasn't able to find a single pixel being off! And the stories you set up in Storybook made checking all of that a breeze, so thank you very much for this!

1. When displaying `topBid` and `minBid` prices, it seems like large numbers or not being abbreviated (eg. in comparison to the "Buy now" price).
2. Speaking about abbreviating large numbers, a while ago we refreshed our internal guidelines on how we want to do that (cc @kdembler). [You can find out all about it over here](https://www.notion.so/Date-time-number-formats-in-Atlas-c5e7c7aec24e4ddfa83104ca4c2e924e), but in a nutshell, you might want to change the decimal separator from a dot `.` to a coma `,` and use spaces as the thousands separator if needed in all numbers across NFT tile & NFT details components.
3. I'm looking at the countdown in the bottom left corner of the Video thumbnail component and I don't quite like how the changing digits affect the width of the pill component they're wrapped within. It's not anything you implemented wrong by any means, and I know this is a Pill component that we're not allowed to edit 100% freely, but I'm wondering... I don't suppose we could add this `font-feature-settings: 'tnum' on, 'lnum' on;` CSS property to the timer in this very pill, could we? What it does is it makes all number characters be of an equal width = no jumping left/right when the countdown is on. Let me know.
4. One adjustment we made with @drillprop to the Video tile component implementation was to have hovering over the Video thumbnail triggering the hover of the details below. Could we make the same sort of adjustment over here as well? The end result would be that when I'm hovering over the video thumbnail, the "NFT details" component below goes into its hover state automatically.
5. Text style used for displaying title of an NFT in the "Small" size does not match the current designs. Right now it's `H200`, but it should be `H300`. This is my fault, as we changed this with @kdembler and @drillprop only last week and I wasn't aware you already implemented this part.

**Comment by toiletgranny** [comment]
Thanks for adding that global hover effect. The only thing left behind is the outline in the Avatar group component, but this one could be super tricky to notice. 😅 
image

---

I know it's outsidre the scope of this issue, but perhaps it's a good opportunity to sneak in this gentle reminder that we're still waiting for that refactor of the Context menu component, @kdembler. Creating/updating the **[List item](https://atlas-storybook-2ywhx1r60-joystream.vercel.app/?path=/story/other-listitem--default)** component was one thing, but the ultimate goal was to have it in the [Context menu](https://atlas-storybook-2ywhx1r60-joystream.vercel.app/?path=/story/overlays-contextmenu--regular) component, along with some other very minor stylistic adjustments, such as changing the background color to be brighter. 

With the current implementation, background of the Context menu component blends with background of the NFT tile component on hover. 

Before | After
-|-
https://user-images.githubusercontent.com/35307309/150322317-4799c90d-4f58-4b17-8983-cf6feb25eb85.mp4 | https://user-images.githubusercontent.com/35307309/150322359-fb4ef371-bf12-4d99-b8c2-f61d3d35156f.mp4

**Review by toiletgranny — CHANGES_REQUESTED** [review]
Thanks for adding that global hover effect. The only thing left behind is the outline in the Avatar group component, but this one could be super tricky to notice. 😅 
image

---

I know it's outsidre the scope of this issue, but perhaps it's a good opportunity to sneak in this gentle reminder that we're still waiting for that refactor of the Context menu component, @kdembler. Creating/updating the **[List item](https://atlas-storybook-2ywhx1r60-joystream.vercel.app/?path=/story/other-listitem--default)** component was one thing, but the ultimate goal was to have it in the [Context menu](https://atlas-storybook-2ywhx1r60-joystream.vercel.app/?path=/story/overlays-contextmenu--regular) component, along with some other very minor stylistic adjustments, such as changing the background color to be brighter. 

With the current implementation, background of the Context menu component blends with background of the NFT tile component on hover. 

Before | After
-|-
https://user-images.githubusercontent.com/35307309/150322317-4799c90d-4f58-4b17-8983-cf6feb25eb85.mp4 | https://user-images.githubusercontent.com/35307309/150322359-fb4ef371-bf12-4d99-b8c2-f61d3d35156f.mp4

**Comment by drillprop** [comment]
> @kdembler. Creating/updating the **[List item](https://atlas-storybook-2ywhx1r60-joystream.vercel.app/?path=/story/other-listitem--default)** component was one thing, but the ultimate goal was to have it in the [Context menu](https://atlas-storybook-2ywhx1r60-joystream.vercel.app/?path=/story/overlays-contextmenu--regular) component, along with some other very minor stylistic adjustments, such as changing the background color to be brighter.

@kdembler Just FYI here's the issue https://github.com/Joystream/atlas/issues/1863

File: src/components/NftTileDetails/NftTileDetails.tsx [hunk 12]
```diff
@@ -173,9 +173,17 @@ export const NftTileDetails: React.FC<NftTileDetailsProps> = ({
           !buyNow && <DetailsContent caption="Status" content="Not for sale" icon={<SvgActionNotForSale />} secondary />
         )
       case 'minBid':
-        return <DetailsContent caption="Min bid" content={minBid || ''} icon={<SvgActionJoyToken />} />
+        return (
+          minBid && (
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.tsx]: This change could be bug-prone. If I set the auction to `minBid` and `minBid` to `0` I'll see: 
image

File: src/components/NftTileDetails/NftTileDetails.tsx [hunk 13]
```diff
@@ -173,9 +173,17 @@ export const NftTileDetails: React.FC<NftTileDetailsProps> = ({
...
+            <DetailsContent caption="Min bid" content={formatNumberShort(minBid)} icon={<SvgActionJoyToken />} />
+          )
+        )
       case 'topBid':
-        return <DetailsContent caption="Top bid" content={topBid || ''} icon={<SvgActionJoyToken />} />
+        return (
+          topBid && (
```
Comment (drillprop) [inline src/components/NftTileDetails/NftTileDetails.tsx]: same as above

File: src/utils/number.ts [hunk 14]
```diff
@@ -15,11 +20,23 @@ export const formatNumberShort = (num: number): string => {
   }
 
   let formattedValue = value.toFixed(1)
+
+  if (Number(formattedValue) >= 1000) {
+    const tempNumber: string[] = []
+    ;[...formattedValue].forEach((value, idx) => {
```
Comment (drillprop) [inline src/utils/number.ts]: Is this possible to avoid this `;` here? This looks a little bit hacky

**Review by drillprop — CHANGES_REQUESTED** [review]

File: src/utils/number.ts (hunk 14 above)
Comment (rafalpawlow) [reply src/utils/number.ts]: Unfortunately this comes from prettier...

Comment (drillprop) [reply src/utils/number.ts]: hmm, we could just do: 
```ts
    const formattedValueAsArray = [...formattedValue]
    formattedValueAsArray.forEach((value, idx) => {
```
right?

**Review by drillprop — APPROVED** [review]

**Comment by kdembler** [comment]
@toiletgranny @drillprop Thank you for bringing this to my attention, this task slipped my mind, moved to `Ready for dev` now so we will tackle next sprint most likely 👍

**Review by toiletgranny — APPROVED** [review]

File: src/utils/number.ts [hunk 15]
```diff
@@ -15,11 +20,24 @@ export const formatNumberShort = (num: number): string => {
...
+    const tempNumber: string[] = []
+    const formattedValueAsArray = [...formattedValue]
+    formattedValueAsArray.forEach((value, idx) => {
+      tempNumber.push(value)
+      if (idx === 0 || (formattedValue.length - idx > 5 && idx % 3 === 0)) {
+        tempNumber.push(' ')
+      }
+    })
+    formattedValue = tempNumber.join('')
+  }
```
Comment (kdembler) [inline src/utils/number.ts]: This looks a bit dark magic-y, what's happening here? When would `Number(formattedValue) >= 1000` happen given the divisions we do above?

**Review by kdembler — CHANGES_REQUESTED** [review]
Screenshot 2022-01-20 at 13 26 38

It seems we broken video tiles along the way. They are all squished together with no margins

File: src/utils/number.ts (hunk 15 above)
Comment (rafalpawlow) [reply src/utils/number.ts]: This function adds whitespace between every thousand. This condition works above 999T.
https://www.notion.so/Date-time-number-formats-in-Atlas-c5e7c7aec24e4ddfa83104ca4c2e924e

Comment (kdembler) [reply src/utils/number.ts]: Is this any different than `value.toLocaleString('pl')`? 🤔 

Btw, we will also be able to handle only up to 9000T with this code, anything above will be over JS integer safe limits so we will need to work with some kind of big numbers

Comment (kdembler) [reply src/utils/number.ts]: Although we can do second part on some different occasion

**Review by kdembler — APPROVED** [review]

**Pull request merged**